File: src/com/blstream/myguide/path/Heap.java

package com.blstream.myguide.path;

public class Heap {

	private Vertex[] mValues;
	private int mLastIndex;

	public Heap(int maxSize) {
		mLastIndex = 0;
		mValues = new Vertex[maxSize + 1];
	}

	public void add(Vertex v) {
		if (mLastIndex + 1 < mValues.length) {
			mLastIndex++;
			int index = mLastIndex;
			mValues[index] = v;
			repairUp(index);
		}
	}

	public Vertex poll() {
		if (mLastIndex < 1) { return null; }
		Vertex v = mValues[1];
		mValues[1] = mValues[mLastIndex];
		mLastIndex--;
		repairDown(1);
		return v;
	}

	public void repairDown(int index) {
		Vertex v = mValues[index];
		while (left(index) <= mLastIndex) {
			int leftIndex = left(index);
			int rightIndex = right(index);
			int smallest = leftIndex;
			if (rightIndex <= mLastIndex && mValues[rightIndex].compareTo(mValues[leftIndex]) == -1) {
				smallest = rightIndex;
			}
			if (mValues[smallest].compareTo(v) != -1) {
				break;
			}
			mValues[index] = mValues[smallest];
			mValues[index].setHeapIndex(index);
			index = smallest;
		}
		mValues[index] = v;
		mValues[index].setHeapIndex(index);
	}

	public void repairUp(int index) {
		if (index > mLastIndex) { return; }
		Vertex v = mValues[index];
		while (parent(index) >= 1 && mValues[parent(index)].compareTo(v) == 1) {
			mValues[index] = mValues[parent(index)];
			mValues[index].setHeapIndex(index);
			index = parent(index);
		}
		mValues[index] = v;
		v.setHeapIndex(index);
	}

	private int parent(int i) {
		return i / 2;
	}

	private int left(int i) {
		return i * 2;
	}

	private int right(int i) {
		return i * 2 + 1;
	}
}

File: MyGuide/Dijkstra/Heap.m
//
// Created by Kamil Lelonek on 5/14/14.
// Copyright (c) 2014 - Open Source (Apache 2.0 license). All rights reserved.
//

#import "Heap.h"

@interface Heap ()

@property (nonatomic) NSUInteger     lastIndex;
@property (nonatomic) NSMutableArray *values;

@end

@implementation Heap

- (id) initWithCapacity: (NSInteger)capacity
{
    self = [super init];
    if (self) {
        _lastIndex = 0;
        _values    = [NSMutableArray arrayWithCapacity: capacity + 1];
    }
    return self;
}

- (void) add: (Vertex *)vertex
{
    if (self.lastIndex + 1 < self.values.count) {
        self.lastIndex++;
        self.values[self.lastIndex] = vertex;
        [self repairUp: self.lastIndex];
    }
}

- (void) repairUp: (NSUInteger)index
{
    if (index > self.lastIndex) return;
    Vertex *vertex = self.values[index];
    while (
            [self parent: index] >= 1
            &&
            [self.values[[self parent: index]] compare: vertex] == NSOrderedAscending) {

        self.values[index] = self.values[[self parent: index]];
        [(Vertex *) self.values[index] setHeapIndex: index];
        index = [self parent: index];
    }
}

- (Vertex *) poll
{
    if (self.lastIndex < 1) return nil;
    Vertex *vertex = self.values[1];
    self.values[1] = [self.values objectAtIndex: self.lastIndex];
    self.lastIndex--;
    [self repairDown: 1];
    return vertex;
}

- (void) repairDown: (NSUInteger)index
{
    Vertex *vertex = self.values[index];
    while ([self left: index] <= self.lastIndex) {
        NSUInteger leftIndex  = [self left: index];
        NSUInteger rightIndex = [self right: index];
        NSUInteger smallest   = leftIndex;
        if (rightIndex <= self.lastIndex && [self.values[rightIndex] compare: self.values[leftIndex]] == NSOrderedDescending) {
            smallest = rightIndex;
        }
        if ([self.values[smallest] compare: vertex] == NSOrderedDescending) break;
        self.values[index] = self.values[smallest];
        [(Vertex *) self.values[index] setHeapIndex: index];
        index = smallest;
    }
    self.values[index] = vertex;
    [(Vertex *) self.values[index] setHeapIndex: index];
}

- (NSUInteger) parent: (NSUInteger)i
{
    return i / 2;
}

- (NSUInteger) left: (NSUInteger)i
{
    return i * 2;
}

- (NSUInteger) right: (NSUInteger)i
{
    return i * 2 + 1;
}

@end
